### PR TITLE
Switch MoE implementation

### PR DESCRIPTION
**Pull request by carmocca**
Benchmark results:

```markdown
# mistralai
eager: 15665.457248687744us
3.9922048 GB
compiled: 15721.099376678467us
3.755603456 GB

# vllm
eager: 49546.36335372925us
6.954663936 GB
compiled: 47361.955642700195us
5.981437952 GB

# dzhulgakov
eager: 15694.758892059326us
4.106371072 GB
compiled: 15854.079723358154us
3.8705152 GB

# huggingface
eager: 19026.272296905518us
4.047395328 GB
compiled: 19472.668170928955us
3.809729024 GB
```

Code:

```python
import time

import torch
import torch.nn as nn


class LLaMAMLP(nn.Module):
    def __init__(self) -> None:
        super().__init__()
        self.fc_1 = nn.Linear(n_embd, intermediate_size, bias=bias)
        self.fc_2 = nn.Linear(n_embd, intermediate_size, bias=bias)
        self.proj = nn.Linear(intermediate_size, n_embd, bias=bias)

    def forward(self, x: torch.Tensor) -> torch.Tensor:
        x_fc_1 = self.fc_1(x)
        x_fc_2 = self.fc_2(x)
        x = torch.nn.functional.silu(x_fc_1) * x_fc_2
        return self.proj(x)


# mistral's
class LLaMAMoE1(nn.Module):
    def __init__(self) -> None:
        super().__init__()
        self.gate = nn.Linear(n_embd, n_expert, bias=False)
        self.experts = nn.ModuleList(LLaMAMLP() for _ in range(n_expert))

    def forward(self, x: torch.Tensor) -> torch.Tensor:
        B, T, C = x.size()  # batch size, sequence length, embedding dimensionality (n_embd)
        x = x.view(-1, C)  # (B*T, C)
        router = self.gate(x)  # (B*T, n_expert)
        probs, indices = torch.topk(router, n_expert_per_token)
        probs = probs.softmax(dim=1, dtype=torch.float).to(dtype=x.dtype)  # (B*T, n_expert_per_token)
        y = torch.zeros_like(x)  # (B*T, C)
        for i, expert in enumerate(self.experts):
            mask = indices == i
            batch_idx, ith_expert = torch.where(mask)
            y[batch_idx] += probs[batch_idx, ith_expert, None] * expert(x[batch_idx])
        return y.view(B, T, C)


# vllm
class LLaMAMoE2(nn.Module):
    def __init__(self) -> None:
        super().__init__()
        self.gate = nn.Linear(n_embd, n_expert, bias=False)
        self.experts = nn.ModuleList(LLaMAMLP() for _ in range(n_expert))

    def forward(self, x: torch.Tensor) -> torch.Tensor:
        B, T, C = x.size()  # batch size, sequence length, embedding dimensionality (n_embd)
        x = x.view(-1, C)  # (B*T, C)
        router = self.gate(x)  # (B*T, n_expert)
        probs, indices = torch.topk(router, n_expert_per_token)
        probs = probs.softmax(dim=1, dtype=torch.float).to(dtype=x.dtype)  # (B*T, n_expert_per_token)
        y = torch.zeros_like(x)  # (B*T, C)
        for i, expert in enumerate(self.experts):
            mask = indices == i
            expert_probs = (probs * mask).sum(dim=-1, keepdim=True)
            y += expert(x) * expert_probs
        return y.view(B, T, C)


# dzhulgakov
class LLaMAMoE3(nn.Module):
    def __init__(self) -> None:
        super().__init__()
        self.gate = nn.Linear(n_embd, n_expert, bias=False)
        self.experts = nn.ModuleList(LLaMAMLP() for _ in range(n_expert))

    def forward(self, x: torch.Tensor) -> torch.Tensor:
        B, T, C = x.size()  # batch size, sequence length, embedding dimensionality (n_embd)
        x = x.view(-1, C)  # (B*T, C)
        router = self.gate(x)  # (B*T, n_expert)
        probs, indices = torch.topk(router, n_expert_per_token, dim=-1)
        probs = probs.softmax(dim=-1, dtype=torch.float).to(dtype=x.dtype)  # (B*T, n_expert_per_token)
        indices = indices.flatten()  # (B*T*n_expert_per_token)
        x = x.repeat_interleave(n_expert_per_token, dim=0)
        y = torch.empty_like(x)  # (B*T*n_expert_per_token, C)
        for i, expert in enumerate(self.experts):
            mask = indices == i
            y[mask] = expert(x[mask])
        y = y.view(*probs.shape, -1) * probs.unsqueeze(-1)  # (B*T, n_expert_per_token, C)
        y = y.sum(dim=1)  # (B*T, C)
        return y.view(B, T, C)


# hf
class LLaMAMoE4(nn.Module):
    def __init__(self) -> None:
        super().__init__()
        self.gate = nn.Linear(n_embd, n_expert, bias=False)
        self.experts = nn.ModuleList(LLaMAMLP() for _ in range(n_expert))

    def forward(self, hidden_states: torch.Tensor) -> torch.Tensor:
        batch_size, sequence_length, hidden_dim = hidden_states.shape
        hidden_states = hidden_states.view(-1, hidden_dim)
        # router_logits: (batch * sequence_length, n_experts)
        router_logits = self.gate(hidden_states)

        routing_weights = torch.nn.functional.softmax(router_logits, dim=1, dtype=torch.float)
        routing_weights, selected_experts = torch.topk(routing_weights, n_expert_per_token, dim=-1)
        routing_weights /= routing_weights.sum(dim=-1, keepdim=True)
        # we cast back to the input dtype
        routing_weights = routing_weights.to(hidden_states.dtype)

        final_hidden_states = torch.zeros(
            (batch_size * sequence_length, hidden_dim), dtype=hidden_states.dtype, device=hidden_states.device
        )

        # One hot encode the selected experts to create an expert mask
        # this will be used to easily index which expert is going to be sollicitated
        expert_mask = torch.nn.functional.one_hot(selected_experts, num_classes=n_expert).permute(2, 1, 0)

        # Loop over all available experts in the model and perform the computation on each expert
        for expert_idx in range(n_expert):
            expert_layer = self.experts[expert_idx]
            idx, top_x = torch.where(expert_mask[expert_idx])

            if top_x.shape[0] == 0:
                continue

            # in torch it is faster to index using lists than torch tensors
            top_x_list = top_x.tolist()
            idx_list = idx.tolist()

            # Index the correct hidden states and compute the expert hidden state for
            # the current expert. We need to make sure to multiply the output hidden
            # states by `routing_weights` on the corresponding tokens (top-1 and top-2)
            current_state = hidden_states[None, top_x_list].reshape(-1, hidden_dim)
            current_hidden_states = expert_layer(current_state) * routing_weights[top_x_list, idx_list, None]

            # However `index_add_` only support torch tensors for indexing so we'll use
            # the `top_x` tensor here.
            final_hidden_states.index_add_(0, top_x, current_hidden_states.to(hidden_states.dtype))
        return final_hidden_states.reshape(batch_size, sequence_length, hidden_dim)


def bench(f, name=None, iters=100, warmup=5):
    for _ in range(warmup):
        f()
    torch.cuda.synchronize()
    begin = time.time()
    for _ in range(iters):
        f()
    torch.cuda.synchronize()
    us_per_iter = (time.time() - begin) * 1e6 / iters
    res = us_per_iter if name is None else f"{name}: {us_per_iter}us"
    print(res)
    return res


n_embd = 4096
n_expert = 8
intermediate_size = 14336
n_expert_per_token = 2
bias = False
torch.set_float32_matmul_precision("high")
device = torch.device("cuda")
B = 1
torch.set_default_dtype(torch.bfloat16)
x = torch.randn(B, 4096, n_embd, device=device)

with device:
    m = LLaMAMoE1()
bench(lambda: m(x), name="eager")
print(torch.cuda.max_memory_allocated() / 1e9)
torch.cuda.reset_peak_memory_stats()

cm = torch.compile(m)
bench(lambda: cm(x), name="compiled")
print(torch.cuda.max_memory_allocated() / 1e9)
```

Additionally, vllm's has a numerical accuracy issue.

This was run on `NVIDIA-SXM4-40GB`, torch nightly.

This PR adopts mistral's.

### CODE REVIEW

**Comment by MF-FOOM** [comment]
thanks for sharing these benchmarks & the code, fascinating!

File: lit_gpt/model.py [hunk 1]
```diff
@@ -299,22 +299,19 @@ def __init__(self, config: Config) -> None:
...
-        probs, indices = torch.topk(router, self.config.n_expert_per_token, dim=-1)
-        probs = probs.softmax(dim=-1)  # (B*T, n_expert_per_token)
-        indices = indices.flatten()  # (B*T*n_expert_per_token)
-        x = x.repeat_interleave(self.config.n_expert_per_token, dim=0)
-        y = torch.empty_like(x)  # (B*T*n_expert_per_token, C)
+        probs, indices = torch.topk(router, self.config.n_expert_per_token)
+        probs = probs.softmax(dim=1, dtype=torch.float).to(dtype=x.dtype)  # (B*T, n_expert_per_token)
+        y = torch.zeros_like(x)  # (B*T, C)
         for i, expert in enumerate(self.experts):
             mask = indices == i
```
Comment (Andrei-Aksionov) [inline lit_gpt/model.py:312]: Have you tried to skip iteration when the mask is empty (all values are False)? 
Does it make any difference in terms of performance?

File: lit_gpt/model.py [hunk 2]
```diff
@@ -299,22 +299,19 @@ def __init__(self, config: Config) -> None:
...
-            y[mask] = expert(x[mask])
-        y = y.view(*probs.shape, -1) * probs.unsqueeze(-1)  # (B*T, n_expert_per_token, C)
-        y = y.sum(dim=1)  # (B*T, C)
+            batch_idx, ith_expert = torch.where(mask)
```
Comment (Andrei-Aksionov) [inline lit_gpt/model.py:313]: I understand that the code is derived from MistralAI's, but technically speaking it's not `batch_idx`, it's a `token_ids`.
Indices have shape of (B*T, n_expert_per_token), so basically it is a list of all tokens and their corresponding expert ids.
Batches have left the chat long time ago, only flattened tokens are left.

**Comment by Andrei-Aksionov** [comment]
When I saw masking ...
https://github.com/Lightning-AI/lit-gpt/blob/74b8df0c3f07fc31d9d1a49e870a1f7955329ad8/lit_gpt/model.py#L312-L313
... I had “Vietnam flashbacks”.

From what I remember, when the size of the tensor is not pre-determined, the CPU does not know what is the shape of the resulting tensor ahead of time. So it has to wait till all GPU operations are finished (CudaStreamSynchronize) to do the correct GPU memory allocation and prepare everything for the next PyTorch ops.

I did a simple profiling from the code in tests (😄), so basically the same model as Mixtral, but only with a single layer to fit it in a A10G.
[mixtral_changes.diff.zip](https://github.com/Lightning-AI/lit-gpt/files/13686935/mixtral_changes.diff.zip)

And from the [trace file](https://github.com/Lightning-AI/lit-gpt/files/13686941/mixtral.json.gz) one can see that the time it takes to do masking takes even longer than to run a single expert.
Screenshot 2023-12-15 at 6 39 19 PM

-----

One of those flashbacks is that it's possible to easily overflow Cuda stream, so by eliminating this bottleneck we can make the forward pass even slower. Nevertheless, I think it's worth to at least try to tackle this issue.

File: lit_gpt/model.py (hunk 1 above)
Comment (carmocca) [reply lit_gpt/model.py:312]: That check would be adding an extra data-dependent operation.

If the mask is all false, `torch.where` will return two empty tensors, and then all the indexing using its result will also result into empty tensors, and forward will get an empty tensor and will return an empty tensor

So I would not add such check.

File: lit_gpt/model.py (hunk 2 above)
Comment (carmocca) [reply lit_gpt/model.py:313]: Makes sense. Would you like to rename it?

**Comment by carmocca** [comment]
Good point @Andrei-Aksionov. Looking at the value to select the expert is unavoidable, however, there might be a smart way to compute the mask, token_idx, and ith_expert for all experts at once outside of the loop, then you would only have a single synchronization point per MoE call.

**Comment by Andrei-Aksionov** [comment]
> This would however tradeoff speed for memory

On a single A10G I only saw increase in memory when the model is compiled and no speedup 🤔.

If to precompute masks/indices ahead of time, we will still see a huge CUDA synchronize at the end of the call as we need to wait till all the experts finish their jobs. In other words, unfortunately, there is no benefit of pre-computation of “selected”.

From the profiing I've noticed two things:
1. Pre-computation with list comprehension takes more time than expected, so a vectorized approach, if feasible, might help.
2. Only one CUDA stream is used. Maybe there is a stream overflow which prevents speedups with precomputed masks/indices. So need to check if there is an overflow and if so, try to maybe create dedicated streams and assign each of experts_per_token to them. But it most likely will be an “academic research” rather than a proper solution. If it was the answer, everyone would use it, I suppose.
